File: src/packages/renderer/templates.py
STATE_TEMPLATE = """
<div class="overlap-wrapper">
  <table class="overlap border-collapse" id="pyagram-state-table">
    <tr>
      <td class="align-top">
        {{ get_frame_html(global_frame) }}
      </td>
      <td class="align-top pl-5">
        {% for object in memory_state %}
          {{ get_object_html(object) }}
        {% endfor %}
      </td>
    </tr>
  </table>
  <svg class="overlap" id="pyagram-svg-canvas" xmlns="http://www.w3.org/2000/svg">
    <defs>
      <marker id="circle" markerWidth="6.5" markerHeight="6.5" refX="5" refY="5">
        <circle cx="5" cy="5" r="1.5" fill="black"/>
      </marker>
      <marker id="arrowhead" markerWidth="6" markerHeight="6" refX="3" refY="5" viewBox="0 0 10 10" orient="auto">
        <path d="M0,0 L10,5 0,10 Z"/>
      </marker>
    </defs>
    <g id="pointers" fill="none" stroke="black" stroke-width="1.5" marker-start="url(#circle)" marker-end="url(#arrowhead)"/>
  </svg>
</div>
"""

ELEMENT_TEMPLATE = """
{% for flag in flags %}
  {{ get_flag_html(flag) }}
{% endfor %}
"""

FLAG_TEMPLATE = """
<div class="pyagram-flag m-3">
  <div class="pyagram-banner {% if is_curr_element %} curr-element {% endif %}">
    <table class="text-center font-family-monospace">
      <tr>
        {% for label, bindings in banner %}
          <td {% if bindings|length > 0 %} colspan="{{ bindings|length }}" {% endif %}>
            {{ label }}
          </td>
        {% endfor %}
      </tr>
      <tr>
        {% for i in range(banner|length) %}
          {% set label = banner[i][0] %}
          {% set bindings = banner[i][1] %}
          {% if bindings|length == 0 %}
            {% if i == 1 %}
              {% if i == banner|length - 1 %}
                <td class="text-left">()</td>
              {% else %}
                <td class="text-left">(</td>
              {% endif %}
            {% elif i == banner|length - 1 %}
              <td class="text-right">)</td>
            {% else %}
              <td class="text-left">,</td>
            {% endif %}
          {% else %}
            {% for binding in bindings %}
              <td class="pyagram-value {% if binding is none %} pyagram-placeholder {% endif %}">
                {% if binding is none %}
                  -
                {% else %}
                  {{ get_reference_html(binding) }}
                {% endif %}
              </td>
            {% endfor %}
          {% endif %}
        {% endfor %}
      </tr>
    </table>
  </div>
  {{ get_element_html(this) }}
  {% if frame is none %}
    <div class="pyagram-placeholder font-family-monospace">...</div>
  {% else %}
    {{ get_frame_html(frame) }}
  {% endif %}
</div>
"""

FRAME_TEMPLATE = """
<div class="pyagram-frame {% if frame_type == 'function' %} mx-3 {% else %} mr-3 {% endif %} my-3 {% if is_curr_element %} curr-element {% endif %}">
  <div class="pyagram-frame-name">
    {% if frame_type == 'function' %}
      {{ name }}
    {% elif frame_type == 'generator' %}
      generator <span class="font-family-monospace">{{ name }}</span>
    {% elif frame_type == 'class' %}
      class <span class="font-family-monospace">{{ name }}</span>
    {% elif frame_type == 'instance' %}
      <span class="font-family-monospace">{{ name }}</span> instance
    {% endif %} {{ get_parent_frame_html(parents, monospace=(frame_type == 'class' or frame_type == 'instance')) }}
  </div>
  <table class="ml-auto mr-0 font-family-monospace">
    {% for key, value in bindings.items() %}
      <tr>
        <td class="text-right">{{ key }}</td>
        <td class="pyagram-value pyagram-frame-value text-left">{{ get_reference_html(value) }}</td>
      </tr>
    {% endfor %}
    {% if return_value is not none %}
      <tr>
        <td class="text-right font-family-sans-serif">
        {% if frame_type == 'generator' %}
          Yield value
        {% else %}
          Return value
        {% endif %}
        </td>
        <td class="pyagram-value pyagram-frame-value text-left">{{ get_reference_html(return_value) }}</td>
      </tr>
    {% endif %}
  </table>
</div>
{{ get_element_html(this) }}
"""

META_REFERENCE_TEMPLATE = """
<span class="pyagram-{{ cls }}">
  {{ text }}
</span>
"""

PLAINTEXT_TEMPLATE = """
{% if monospace %}
  <span class="font-family-monospace">{{ text }}</span>
{% else %}
  {{ text }}
{% endif %}
"""

POINTER_TEMPLATE = """
<span class="pyagram-placeholder pyagram-reference font-family-monospace reference-{{ id }}">
  -
</span>
"""

OBJECT_TEMPLATE = """
<div id="object-{{ id }}" class="pyagram-object m-3">
  {{ get_object_body_html(object) }}
</div>
"""

FUNCTION_TEMPLATE = """
{% if is_gen_func %}
  generator function
{% else %}
  function
{% endif %}
<span class="ml-2 font-family-monospace">
  {% if lambda_id is none %}
    {{ name }}
  {% else %}
    {{ get_lambda_html(lambda_id) }}
  {% endif %}
  (
  {% for i in range(parameters|length) %}
    {% set parameter = parameters[i] %}
    {{ get_parameter_html(parameter) }}
    {% if i < parameters|length - 1 %}, {% endif %}
  {% endfor %}
  )
</span>
<div>
  {{ get_parent_frame_html([parent]) }}
</div>
"""

BUILTIN_FUNCTION_TEMPLATE = """
function
<span class="ml-2 font-family-monospace">
  {{ name }}(...)
</span>
"""

LAMBDA_TEMPLATE = """
&#955;
<sub>
  {{ lineno }}
  {% if not single %}
    #{{ number }}
  {% endif %}
</sub>
"""

ORDERED_COLLECTION_TEMPLATE = """
{% if elements|length == 0 %}
  empty {{ type }}
{% else %}
  <div class="d-flex flex-row align-items-center">
    <div>{{type}}</div>
    <table class="pyagram-ordered-collection border-collapse ml-1 font-family-monospace" rules="cols">
    <tr>
      {% for element in elements %}
        <td class="pyagram-collection-element px-2">{{ get_reference_html(element) }}</td>
      {% endfor %}
    </tr>
  </table>
  </div>
{% endif %}
"""

UNORDERED_COLLECTION_TEMPLATE = """
{% if elements|length == 0 %}
  empty {{ type }}
{% else %}
  <div class="d-flex flex-row align-items-center">
    <div>{{type}}</div>
    <table class="pyagram-unordered-collection ml-1 font-family-monospace">
    <tr>
      {% for element in elements %}
        <td class="pyagram-collection-element px-2">{{ get_reference_html(element) }}</td>
      {% endfor %}
    </tr>
  </table>
  </div>
{% endif %}
"""

MAPPING_TEMPLATE = """
{% if items|length == 0 %}
  empty {{ type }}
{% else %}
  <div class="d-flex flex-row align-items-center">
    <div>{{type}}</div>
    <table class="pyagram-mapping border-collapse ml-1 font-family-monospace" rules="cols">
    <tr>
      {% for item in items %}
        <td class="pyagram-mapping-key px-2 text-center">{{ get_reference_html(item[0]) }}</td>
      {% endfor %}
    </tr>
    <tr>
      {% for item in items %}
        <td class="pyagram-mapping-value px-2 text-center">{{ get_reference_html(item[1]) }}</td>
      {% endfor %}
    </tr>
  </table>
  </div>
{% endif %}
"""

ITERATOR_TEMPLATE = """
iterator over <span class="pyagram-value">{{ get_reference_html(object) }}</span>
{% if annotation is not none %}
  <span> {{ annotation }}</span>
{% endif %}
<div>[next index: {{ index }}]</div>
"""

# TODO: Reconsider how you display these.
OBJECT_REPR_TEMPLATE = """
<div class="font-family-monospace">
  {{ repr }}
</div>
"""

PARENT_FRAME_TEMPLATE = """
{% if parents|length == 0 %}
{% elif parents|length == 1 %}
  [parent: {{ parents[0] }}]
{% else %}
  [parents: {% for i in range(parents|length - 1) %}{{ parents[i] }}, {% endfor %}{{ parents[-1] }}]
{% endif %}
"""

PARAMETER_TEMPLATE = """
{{ name }}
{% if default is not none %}
  =<span class="pyagram-value">{{ get_reference_html(default) }}</span>
{% endif %}
"""

PRINT_TEMPLATE = """
{% for line in print_output %}
  <div class="print-output {% if is_exception %} pyagram-exception {% endif %} font-family-monospace">
    {{ line }}
  </div>
{% endfor %}
"""

File: src/scripts/templates.js
function compile(template) {
    return Handlebars.compile(
        template.split('\n').map(function(line) {
            line = line.replace(/^\s+/, '');
            line = line.replace(/\s+$/, '');
            return line;
        }).join(''),
        {
            noEscape: true,
            assumeObjects: true,
        },
    );
}

export const ESCAPE = Handlebars.compile(`{{this}}`);

export const STACK_TEMPLATE = compile(`
<div class="font-family-monospace">
  {{decodeFrameSnapshot this}}
</div>
`);

export const ELEMENT_TEMPLATE = compile(`
{{#each flags}}
  {{decodeFlagSnapshot this}}
{{/each}}
`);

export const FLAG_TEMPLATE = compile(`
<div class="pyagram-flag m-3">
  <div class="pyagram-banner mr-3 {{#if is_curr_element}} curr-element {{/if}}">
    <table class="text-center">
      {{#if (isEqual type 'call')}}
        <tr>
          {{#each banner}}
            <td colspan="{{n_cols}}">
              {{escape code}}
            </td>
            {{#if @first}}
              {{#if @last}}
                <td>()</td>
              {{else}}
                <td>(</td>
              {{/if}}
            {{else}}
              {{#if @last}}
                <td>)</td>
              {{else}}
                <td>,</td>
              {{/if}}
            {{/if}}
          {{/each}}
        </tr>
        <tr>
          {{#each banner}}
            {{#if (isNull bindings)}}
              <td class="pyagram-value pyagram-placeholder">
                -
              </td>
            {{else if (isEmpty bindings)}}
              <td></td>
            {{else}}
              {{#each bindings}}
                {{#unless (isNull key)}}
                  <td class="banner-text">{{decodeReferenceSnapshot key}}=</td>
                {{/unless}}
                <td class="pyagram-value">
                  {{decodeReferenceSnapshot value}}
                </td>
                {{#unless @last}}
                  <td class="banner-text">,</td>
                {{/unless}}
              {{/each}}
            {{/if}}
            {{#if @first}}
              {{#if @last}}
                <td>()</td>
              {{else}}
                <td>(</td>
              {{/if}}
            {{else}}
              {{#if @last}}
                <td>)</td>
              {{else}}
                {{#if (isNull bindings)}}
                  <td>,</td>
                {{else if (isEmpty bindings)}}
                  <td></td>
                {{else}}
                  <td>,</td>
                {{/if}}
              {{/if}}
            {{/if}}
          {{/each}}
        </tr>
      {{else if (isEqual type 'comp')}}
        <tr>
          {{#each banner}}
            <td>
              {{escape code}}
            </td>
          {{/each}}
        </tr>
      {{/if}}
    </table>
  </div>
  {{decodeElementSnapshot this}}
  {{#if (isNull frame)}}
    <div class="pyagram-placeholder">
      -
    </div>
  {{else}}
    {{decodeFrameSnapshot frame}}
  {{/if}}
</div>
`);

export const FRAME_TEMPLATE = compile(`
<div class="pyagram-frame {{#if (isEqual type 'function')}} m-3 {{/if}} {{#if is_curr_element}} curr-element {{/if}}">
  <div class="pyagram-frame-name">
    {{#if (isEqual type 'function')}}
      <span class="font-family-sans-serif">
        {{escape name}}{{#unless (isNull parent)}} [parent: {{escape parent}}]{{/unless}}
      </span>
    {{else if (isEqual type 'generator')}}
      <span class="font-family-sans-serif">
        {{escape name}} [parent: {{escape parent}}]
      </span>
    {{else if (isEqual type 'class')}}
      <span class="font-family-sans-serif">
        class
        {{space 1}}
      </span>
      {{escape name}}
      {{#if (isNull parents)}}
        <span class="font-family-sans-serif">
          {{space 1}}
          [parent: <span class="font-family-monospace">{{decodeUnknownSnapshot this}}</span>]
        </span>
      {{else if (isEmpty parents)}}
      {{else}}
        <span class="font-family-sans-serif">
          {{space 1}}
          [parent{{#unless (isEqual parents.length 1)}}s{{/unless}}:
          {{space 1}}
          {{#each parents}}
            class <span class="font-family-monospace">{{escape this}}</span>
            {{#unless @last}}, {{/unless}}
          {{/each}}
          ]
        </span>
      {{/if}}
    {{else if (isEqual type 'instance')}}
      {{escape name}}
      <span class="font-family-sans-serif">
        {{space 1}}
        instance [parent: class <span class="font-family-monospace">{{escape parent}}</span>]
      </span>
    {{/if}}
  </div>
  <table class="ml-auto mr-0">
    {{#each bindings}}
      <tr>
        <td class="text-right">
          {{decodeReferenceSnapshot key}}
        </td>
        <td class="text-left pyagram-value pyagram-frame-value" {{#if (isEqual ../type 'generator')}}{{#unless (isNull ../from)}} colspan="3" {{/unless}}{{/if}}>
          {{decodeReferenceSnapshot value}}
        </td>
      </tr>
    {{/each}}
    {{#unless (isNull return_value)}}
      <tr>
        <td class="text-right font-family-sans-serif">
          {{#if (isEqual type 'generator')}}
            Yield:
          {{else}}
            Return:
          {{/if}}
        </td>
        <td class="text-left pyagram-value pyagram-frame-value">
          {{decodeReferenceSnapshot return_value}}
        </td>
        {{#if (isEqual type 'generator')}}
          {{#unless (isNull from)}}
            <td class="text-left font-family-sans-serif">
              from
            </td>
            <td class="text-left pyagram-value">
              {{decodeReferenceSnapshot from}}
            </td>
          {{/unless}}
        {{/if}}
      </tr>
    {{/unless}}
  </table>
  {{#if (isEqual type 'class')}}
    {{#if bltn}}
      <div class="text-center">
        ...
      </div>
    {{/if}}
  {{/if}}
</div>
{{#if (isEqual type 'function')}}
  {{decodeElementSnapshot this}}
{{/if}}
`);

export const HEAP_TEMPLATE_TEXTPOINTERS_T = compile(`
<table class="border-collapse my-2 font-family-monospace">
  {{#each this}}
    <tr>
      <td class="px-2 py-0 font-family-sans-serif">
        @{{decodeObjectIdSnapshot id}})
      </td>
      <td class="p-0">
        <div class="pyagram-object my-2 mr-3">
          {{decodeEncodedObjectSnapshot object}}
        </div>
      </td>
    </tr>
  {{/each}}
</table>
`);

export const HEAP_TEMPLATE_TEXTPOINTERS_F = compile(`
<table class="border-collapse my-2 font-family-monospace">
  {{#each this}}
    <tr>
      <td class="p-0">
        <div class="pyagram-object my-2 mr-3" id="object-{{decodeObjectIdSnapshot id}}">
          {{decodeEncodedObjectSnapshot object}}
        </div>
      </td>
    </tr>
  {{/each}}
</table>
`);

export const OMITTED_TEMPLATE = compile(`
<span title="This value cannot be displayed without compromising the pyagram's accuracy.">...</span>
`)

export const UNKNOWN_TEMPLATE = compile(`
<span class="pyagram-unknown">(?)</span>
`);

export const PRIMITIVE_TEMPLATE = compile(`
{{escape this}}
`);

export const REFERENT_TEMPLATE_TEXTPOINTERS_T = compile(`
<span class="font-family-sans-serif">@{{decodeObjectIdSnapshot this}}</span>
`);

export const REFERENT_TEMPLATE_TEXTPOINTERS_F = compile(`
<span class="pyagram-placeholder pyagram-pointer reference-{{decodeObjectIdSnapshot this}}">-</span>
`);

export const OBJECT_ID_TEMPLATE = compile(`
{{this}}
`);

export const FUNCTION_TEMPLATE = compile(`
<span class="font-family-sans-serif">
  {{#if is_gen_func}}
    generator function
  {{else}}
    function
  {{/if}}
  {{space 1}}
</span>
{{#if (isNull lambda_id)}}
  {{escape name}}
{{else}}
  {{#with lambda_id}}
    &#955;<sub>{{lineno}}{{#unless single}}#{{number}}{{/unless}}</sub>
  {{/with}}
{{/if}}
(
{{#each parameters}}
  {{escape name}}
  {{#unless (isNull default)}}
    =<span class="pyagram-value">{{decodeReferenceSnapshot default}}</span>
  {{/unless}}
  {{#unless @last}}, {{/unless}}
{{/each}}
)
<div class="font-family-sans-serif">
  [parent: {{escape parent}}]
</div>
`);

export const METHOD_TEMPLATE = compile(`
<span class="font-family-sans-serif">
  method
  {{space 1}}
</span>
<span class="pyagram-value">
  {{decodeReferenceSnapshot function}}
</span>
<span class="font-family-sans-serif">
  {{space 1}}
  bound to
  {{space 1}}
</span>
<span class="pyagram-value">
  {{decodeReferenceSnapshot instance}}
</span>
`);

export const BUILTIN_TEMPLATE = compile(`
<span class="font-family-sans-serif">
  {{#if (isNull instance)}}
    function
  {{else}}
    method
  {{/if}}
  {{space 1}}
</span>
{{escape name}}(...)
{{#unless (isNull instance)}}
  <span class="font-family-sans-serif">
    {{space 1}}
    bound to
    {{space 1}}
  </span>
  <span class="pyagram-value">
    {{decodeReferenceSnapshot instance}}
  </span>
{{/unless}}
`);

export const ORDERED_COLLECTION_TEMPLATE = compile(`
{{#if (isEmpty elements)}}
  <div class="font-family-sans-serif">
    empty {{escape type}}
  </div>
{{else}}
  <div class="d-flex flex-row align-items-center">
    <div class="font-family-sans-serif">
      {{escape type}}
    </div>
    <table class="pyagram-ordered-collection border-collapse ml-1" rules="cols">
      <tr>
        {{#each elements}}
          <td class="pyagram-collection-element px-2">{{decodeReferenceSnapshot this}}</td>
        {{/each}}
      </tr>
    </table>
  </div>
{{/if}}
`);

export const UNORDERED_COLLECTION_TEMPLATE = compile(`
{{#if (isEmpty elements)}}
  <div class="font-family-sans-serif">
    empty {{escape type}}
  </div>
{{else}}
  <div class="d-flex flex-row align-items-center">
    <div class="font-family-sans-serif">
      {{escape type}}
    </div>
    <table class="pyagram-unordered-collection ml-1">
      <tr>
        {{#each elements}}
          <td class="pyagram-collection-element px-2">{{decodeReferenceSnapshot this}}</td>
        {{/each}}
      </tr>
    </table>
  </div>
{{/if}}
`);

export const MAPPING_TEMPLATE = compile(`
{{#if (isEmpty items)}}
  <div class="font-family-sans-serif">
    empty {{escape type}}
  </div>
{{else}}
  <div class="d-flex flex-row align-items-center">
    <div class="font-family-sans-serif">
      {{escape type}}
    </div>
    <table class="pyagram-mapping border-collapse ml-1" rules="cols">
      <tr>
        {{#each items}}
          <td class="pyagram-mapping-key px-2 text-center">{{decodeReferenceSnapshot key}}</td>
        {{/each}}
      </tr>
      <tr>
        {{#each items}}
          <td class="pyagram-mapping-value px-2 text-center">{{decodeReferenceSnapshot value}}</td>
        {{/each}}
      </tr>
    </table>
  </div>
{{/if}}
`);

export const ITERATOR_TEMPLATE = compile(`
{{#if (isNull this)}}
  <div class="font-family-sans-serif">
    empty iterator
  </div>
{{else}}
  <span class="font-family-sans-serif">
    iterator over
    {{space 1}}
  </span>
  <span class="pyagram-value">
    {{decodeReferenceSnapshot object}}
  </span>
  {{#unless (isNull annotation)}}
    <span class="font-family-sans-serif">
      {{space 1}}
      {{escape annotation}}
    </span>
  {{/unless}}
  <div class="font-family-sans-serif">
    [next index: {{index}}]
  </div>
{{/if}}
`);

export const GENERATOR_TEMPLATE = compile(`
<span class="font-family-sans-serif">
  generator
  {{space 1}}
</span>
{{escape name}}
{{decodeFrameSnapshot frame}}
`);

export const RANGE_TEMPLATE = compile(`
<span class="font-family-sans-serif">
  range
  {{space 1}}
</span>
[{{decodeReferenceSnapshot start}}, {{decodeReferenceSnapshot stop}})
<span class="font-family-sans-serif">
  {{space 1}}
  with step size
  {{space 1}}
</span>
{{decodeReferenceSnapshot step}}
`)

export const SLICE_TEMPLATE = compile(`
<span class="font-family-sans-serif">
  slice
  {{space 1}}
</span>
[
{{#unless (isNull start)}}{{decodeReferenceSnapshot start}}{{/unless}}
:
{{#unless (isNull stop)}}{{decodeReferenceSnapshot stop}}{{/unless}}
:
{{#unless (isNull stop)}}{{decodeReferenceSnapshot step}}{{/unless}}
]
`)

export const OTHER_TEMPLATE = compile(`
{{escape this}}
`);

export const EXCEPTION_TEMPLATE = compile(`
{{#unless (isNull this)}}
  <div class="px-3 py-2 pyagram-readout font-family-monospace">
    {{escape this}}
  </div>
{{/unless}}
`);

export const PRINT_OUTPUT_TEMPLATE = compile(`
<div class="pyagram-readout font-family-monospace">
  {{escape this}}
</div>
`);

export const ERROR_TEMPLATE = compile(`
<div class="p-3 pyagram-readout font-family-monospace">
  <div>
    {{escape type}}{{#unless (isNull lineno)}} (line {{lineno}}){{/unless}}:
  </div>
  {{#if (isEqual encoding 'pyagram')}}
    {{decodePyagramError data}}
  {{else if (isEqual encoding 'syntax')}}
    {{decodeSyntaxError data}}
  {{/if}}
</div>
`);

export const PYAGRAM_ERROR_TEMPLATE = compile(`
{{message}}
`)

export const SYNTAX_ERROR_TEMPLATE = compile(`
<div>
  {{escape code}}
</div>
<div>
  {{space offset}}^
</div>
`)
